### PR TITLE
Upgrade `FrozenTrial._validate()`

### PR DESCRIPTION
**Pull request by eukaryo**
## Motivation

More clearly define specifications of trial and increase the strictness of the software.

## Description of the changes

Added codes to check:
- _values is None if state is PRUNED or FAIL
- _values does not include NaN if state is COMPLETE

### CODE REVIEW

**Review by HideakiImamura — DISMISSED** [review]
LGTM.

File: optuna/trial/_frozen.py [hunk 1]
```diff
@@ -321,8 +322,15 @@ def _validate(self) -> None:
                     "`datetime_complete` is supposed to be None for an unfinished trial."
                 )
 
+        if self.state == TrialState.PRUNED and self._values is not None:
+            raise ValueError("`value` is supposed not to be set for a pruned trial.")
```
Comment (contramundum53) [inline optuna/trial/_frozen.py:326]:
```suggestion
            raise ValueError("`value` is not supposed to be set for a pruned trial.")
```
[nit] Fix grammatical error

File: optuna/trial/_frozen.py [hunk 2]
```diff
@@ -321,8 +322,15 @@ def _validate(self) -> None:
...
+        if self.state == TrialState.FAIL and self._values is not None:
+            raise ValueError("`value` is supposed not to be set for a failed trial.")
```
Comment (contramundum53) [inline optuna/trial/_frozen.py:328]:
```suggestion
        if self.state in (TrialState.PRUNED, TrialState.FAIL) and self._values is not None:
            raise ValueError(f"values should be None for a trial with state {self.state}, but got {self._values}.")
```
[nit] Make the code readable and easier to debug for users.

File: optuna/trial/_frozen.py [hunk 3]
```diff
@@ -321,8 +322,15 @@ def _validate(self) -> None:
...
         if self.state == TrialState.COMPLETE and self._values is None:
             raise ValueError("`value` is supposed to be set for a complete trial.")
```
Comment (contramundum53) [inline optuna/trial/_frozen.py:330]:
```suggestion
            raise ValueError("values should be set for a complete trial.")
```
[nit] Just making the format the same.

File: optuna/trial/_frozen.py [hunk 4]
```diff
@@ -321,8 +322,15 @@ def _validate(self) -> None:
...
+        self_values: List[float] = self._values if self._values is not None else []  # for mypy
+        if self.state == TrialState.COMPLETE and any([math.isnan(x) for x in self_values]):
+            raise ValueError("`value` is supposed not to be NaN for a complete trial.")
```
Comment (contramundum53) [inline optuna/trial/_frozen.py:333]:
```suggestion
        if self.state == TrialState.COMPLETE:
            if self._values is None:
                raise ValueError("values should be set for a complete trial.")
            elif any(math.isnan(x) for x in self._values):
                raise ValueError("values should not contain NaN.")
```
A better way to silence mypy

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Pull request merged**